packages/@atjson/document/src/index.ts

```diff
@@ -472,11 +517,110 @@ export default class Document {
+    let lastEnd;
+    for (let i = 0; i < mergedRanges.length - 1; i++) {
+      newContent += this.content.slice(
+        mergedRanges[i].end,
+        mergedRanges[i + 1].start
+      );
+      lastEnd = mergedRanges[i + 1].end;
+    }
+
+    this.content = newContent + this.content.slice(lastEnd);
+
+    /**
+     * for adjusting annotations, we need to handle the ranges backwards.
+     * Conveniently they are already sorted and non-overlapping, so we can simply walk the list backwards
+     */
+
+    for (let i = mergedRanges.length - 1; i >= 0; i--) {
+      let change = mergedRanges[i];
+
+      for (let annotation of this.annotations) {
+        let length = change.end - change.start;
+
+        // We're deleting after the annotation, nothing needed to be done.
+        //    [   ]
+        // -----------*---*---
+        if (annotation.end < change.start) {
+          continue;
+        }
+
+        // If the annotation is wholly *after* the deleted text, just move
+        // everything.
+        //           [       ]
+        // --*---*-------------
+        if (change.end < annotation.start) {
+          annotation.start -= length;
+          annotation.end -= length;
+        } else {
+          if (change.end < annotation.end) {
+            // Annotation spans the whole deleted text, so just truncate the end of
+            // the annotation (shrink from the right).
+            //   [             ]
+            // ------*------*---------
+            if (change.start > annotation.start) {
+              annotation.end -= length;
+
+              // Annotation occurs within the deleted text, affecting both start and
+              // end of the annotation, but by only part of the deleted text length.
+              //         [         ]
+              // ---*---------*------------
+            } else if (change.start <= annotation.start) {
+              annotation.start -= annotation.start - change.start;
+              annotation.end -= length;
+            }
+          } else if (change.end >= annotation.end) {
+            //             [  ]
+            //          [     ]
+            //          [         ]
+            //              [     ]
+            //    ------*---------*--------
+            if (change.start <= annotation.start) {
+              annotation.start = change.start;
+              annotation.end = change.start;
+
+              //       [        ]
+              //    ------*---------*--------
+            } else if (change.start > annotation.start) {
+              annotation.end = change.start;
+            }
+          }
+        }
+      }
+    }
+
+    this.triggerChange();
+  }
+
   canonical() {
     let canonicalDoc = this.clone();
+    let ranges: Array<{ start: number; end: number }> = [];
+
     canonicalDoc.where({ type: "-atjson-parse-token" }).update(a => {
-      canonicalDoc.deleteText(a.start, a.end);
+      ranges.push({ start: a.start, end: a.end });
```

Comment:
@colinarobinson re: slack convo, just hiding `ranges.push()` inside `doc.deleteText()` and then calling `deleteTextRanges` before the next `where` call (or the next access of `content`, etc ... It's the "when does `deleteTextRanges` need to be called, and how?" question that's a bit unclear to me, but if there's a nice way of doing it, it would 'guarantee' better performance across the board.

(totally an enhancement, though, by no means blocking or necessary to ship this)

Reply:
cool! yeah in general I like laziness but yeah the question of when exactly it needs to be 'flushed' feels a little hairy to tackle in this pr